relay-system/src/service.rs

```diff
@@ -237,6 +239,232 @@ impl MessageResponse for NoResponse {
+    Pending(oneshot::Receiver<InitialResponse<T>>),
+    /// The request is attached to a [`BroadcastChannel`].
+    Attached(Shared<oneshot::Receiver<T>>),
+}
+
+/// The request when sending an asynchronous message to a service.
+///
+/// This is returned from [`Addr::send`] when the message responds asynchronously through
+/// [`BroadcastResponse`]. It is a future that should be awaited. The message still runs to
+/// completion if this future is dropped.
+///
+/// # Panics
+///
+/// This future is not fused and panics if it is polled again after it has resolved.
+pub struct BroadcastRequest<T>(BroadcastState<T>)
+where
+    T: Clone;
+
+impl<T: Clone> Future for BroadcastRequest<T> {
+    type Output = Result<T, SendError>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        Poll::Ready(loop {
+            match self.0 {
+                BroadcastState::Pending(ref mut pending) => {
+                    match futures::ready!(Pin::new(pending).poll(cx)) {
+                        Ok(InitialResponse::Ready(value)) => break Ok(value),
+                        Ok(InitialResponse::Poll(shared)) => {
+                            self.0 = BroadcastState::Attached(shared)
+                        }
+                        Err(_) => break Err(SendError),
+                    }
+                }
+                BroadcastState::Attached(ref mut shared) => {
+                    match futures::ready!(Pin::new(shared).poll(cx)) {
+                        Ok(value) => break Ok(value),
+                        Err(_) => break Err(SendError),
+                    }
+                }
+            }
+        })
+    }
+}
+
+/// A channel that broadcasts values to attached [senders](BroadcastSender).
+///
+/// This is part of the [`BroadcastResponse`] message behavior to efficiently send delayed responses
+/// to a large number of senders. All requests that are attached to this channel via their senders
+/// resolve with the same value.
+///
+/// # Example
+///
+/// ```
+/// todo!()
+/// ```
+#[derive(Debug)]
+pub struct BroadcastChannel<T>
+where
+    T: Clone,
+{
+    tx: oneshot::Sender<T>,
+    rx: Shared<oneshot::Receiver<T>>,
+}
+
+impl<T: Clone> BroadcastChannel<T> {
+    /// Creates a standalone channel.
+    ///
+    /// Use [`attach`](Self::attach) to add senders to this channel. Alternatively, create a channel
+    /// with [`BroadcastSender::into_channel`].
+    pub fn new() -> Self {
+        let (tx, rx) = oneshot::channel();
+        Self {
+            tx,
+            rx: rx.shared(),
+        }
+    }
+
+    /// Attaches a sender of another message to this channel to receive the same value.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// todo!()
+    /// ```
+    pub fn attach(&mut self, sender: BroadcastSender<T>) {
+        sender.0.send(InitialResponse::Poll(self.rx.clone())).ok();
+    }
+
+    /// Send a value to all attached senders and close the channel.
```

Comment:
```suggestion
    /// Sends a value to all attached senders and close the channel.
```

If we want to keep the docs consistent.